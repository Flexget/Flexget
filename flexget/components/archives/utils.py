"""
Utilities for handling RAR and ZIP archives

Provides wrapper archive and exception classes to simplify
archive extraction
"""


import os
import shutil
import zipfile

from loguru import logger

try:
    import rarfile
except ImportError:
    rarfile = None

logger = logger.bind(name='archive')


class ArchiveError(Exception):
    """Base exception for archive"""

    pass


class NeedRarFile(ArchiveError):
    """Exception to be raised when rarfile module is missing"""

    pass


class BadArchive(ArchiveError):
    """Wrapper exception for BadZipFile and BadRarFile"""

    pass


class NeedFirstVolume(ArchiveError):
    """Wrapper exception for rarfile.NeedFirstVolume"""

    pass


class PathError(ArchiveError):
    """Exception to be raised when an archive file doesn't exist"""

    pass


class FSError(ArchiveError):
    """Exception to be raised on OS/IO exceptions"""

    pass


class FileAlreadyExists(ArchiveError):
    """Exception to be raised when destination file already exists"""

    pass


def rarfile_set_tool_path(config):
    """
    Manually set the path of unrar executable if it can't be resolved from the
    PATH environment variable
    """
    unrar_tool = config['unrar_tool']

    if unrar_tool:
        if not rarfile:
            logger.error('rar_tool specified with no rarfile module installed.')
        else:
            rarfile.UNRAR_TOOL = unrar_tool
            logger.debug('Set RarFile.unrar_tool to: {}', unrar_tool)


def rarfile_set_path_sep(separator):
    """
    Set the path separator on rarfile module
    """
    if rarfile:
        rarfile.PATH_SEP = separator


def makepath(path):
    """Make directories as needed"""
    if not os.path.exists(path):
        logger.debug('Creating path: {}', path)
        os.makedirs(path)


class Archive:
    """
    Base archive class. Assumes an interface similar to
    zipfile.ZipFile or rarfile.RarFile
    """

    def __init__(self, archive_object, path):
        self.path = path

        self.archive = archive_object(self.path)

    def close(self):
        """Release open resources."""
        self.archive.close()

    def delete(self):
        """Delete the volumes that make up this archive"""
        volumes = self.volumes()
        self.close()

        try:
            for volume in volumes:
                os.remove(volume)
                logger.verbose('Deleted archive: {}', volume)
        except OSError as error:
            raise FSError(error)

    def volumes(self):
        """Returns the list of volumes that comprise this archive"""
        return [self.path]

    def infolist(self):
        """Returns a list of info objects describing the contents of this archive"""
        infolist = []

        for info in self.archive.infolist():
            try:
                archive_info = ArchiveInfo(info)
                infolist.append(archive_info)
            except ValueError as e:
                logger.debug(e)

        return infolist

    def open(self, member):
        """Returns file-like object from where the data of a member file can be read."""
        return self.archive.open(member)

    def extract_file(self, member, destination):
        """Extract a member file to the specified destination"""
        try:
            with self.open(member) as source:
                with open(destination, 'wb') as target:
                    shutil.copyfileobj(source, target)
        except OSError as error:
            raise FSError(error)


class RarArchive(Archive):
    """
    Wrapper class for rarfile.RarFile
    """

    def __init__(self, path):
        RarArchive.check_import()

        try:
            super().__init__(rarfile.RarFile, path)
        except rarfile.BadRarFile as error:
            raise BadArchive(error)
        except rarfile.NeedFirstVolume as error:
            raise NeedFirstVolume(error)
        except rarfile.Error as error:
            raise ArchiveError(error)

    def volumes(self):
        """Returns the list of volumes that comprise this archive"""
        return self.archive.volumelist()

    def open(self, member):
        """Returns file-like object from where the data of a member file can be read."""
        try:
            return super().open(member)
        except rarfile.Error as error:
            raise ArchiveError(error)

    @staticmethod
    def check_import():
        if not rarfile:
            raise NeedRarFile('Python module rarfile needed to handle RAR archives')


class ZipArchive(Archive):
    """
    Wrapper class for zipfile.ZipFile
    """

    def __init__(self, path):
        try:
            super().__init__(zipfile.ZipFile, path)
        except zipfile.BadZipfile as error:
            raise BadArchive(error)

    def open(self, member):
        """Returns file-like object from where the data of a member file can be read."""
        try:
            return super().open(member)
        except zipfile.BadZipfile as error:
            raise ArchiveError(error)


class ArchiveInfo:
    """Wrapper class for  archive info objects"""

    def __init__(self, info):
        self.info = info
        self.path = info.filename
        self.filename = os.path.basename(self.path)

        if self._is_dir():
            raise ValueError('Appears to be a directory: %s' % self.path)

    def _is_dir(self):
        """Indicates if info object looks to be a directory"""

        if hasattr(self.info, 'isdir'):
            return self.info.isdir()
        else:
            return not self.filename

    def extract(self, archive, destination):
        """Extract ArchiveInfo object to the specified destination"""
        dest_dir = os.path.dirname(destination)

        if os.path.exists(destination):
            raise FileAlreadyExists('File already exists: %s' % destination)

        logger.debug('Creating path: {}', dest_dir)
        makepath(dest_dir)

        try:
            archive.extract_file(self.info, destination)
            logger.verbose('Extracted: {} to {}', self.path, destination)
        except Exception as error:
            if os.path.exists(destination):
                logger.debug('Cleaning up partially extracted file: {}', destination)
                os.remove(destination)

            raise error


def open_archive(archive_path):
    """
    Returns the appropriate archive object
    """

    archive = None

    if not os.path.exists(archive_path):
        raise PathError('Path doesn\'t exist')

    if zipfile.is_zipfile(archive_path):
        archive = ZipArchive(archive_path)
        logger.debug('Successfully opened ZIP: {}', archive_path)
    elif rarfile and rarfile.is_rarfile(archive_path):
        archive = RarArchive(archive_path)
        logger.debug('Successfully opened RAR: {}', archive_path)
    else:
        if not rarfile:
            logger.warning('Rarfile module not installed; unable to handle RAR archives.')

    return archive


def is_archive(path):
    """
    Attempts to open an entry as an archive; returns True on success, False on failure.
    """

    archive = None

    try:
        archive = open_archive(path)
        if archive:
            archive.close()
            return True
    except (OSError, ArchiveError) as error:
        logger.debug('Failed to open file as archive: {} ({})', path, error)

    return False
