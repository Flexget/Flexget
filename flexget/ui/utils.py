def pretty_date(time=False):
    """
    Get a datetime object or a int() Epoch timestamp and return a
    pretty string like 'an hour ago', 'Yesterday', '3 months ago',
    'just now', etc
    """
    from datetime import datetime
    now = datetime.now()
    diff = now - datetime.fromtimestamp(time)
    second_diff = diff.seconds
    day_diff = diff.days

    if day_diff < 0:
        return ''

    if day_diff == 0:
        if second_diff < 10:
            return "just now"
        if second_diff < 60:
            return str(second_diff) + " seconds ago"
        if second_diff < 120:
            return "a minute ago"
        if second_diff < 3600:
            return str(second_diff / 60) + " minutes ago"
        if second_diff < 7200:
            return "an hour ago"
        if second_diff < 86400:
            return str(second_diff / 3600) + " hours ago"
    if day_diff == 1:
        return "Yesterday"
    if day_diff < 7:
        return str(day_diff) + " days ago"
    if day_diff < 31:
        return str(day_diff / 7) + " weeks ago"
    if day_diff < 365:
        return str(day_diff / 30) + " months ago"
    return str(day_diff / 365) + " years ago"
