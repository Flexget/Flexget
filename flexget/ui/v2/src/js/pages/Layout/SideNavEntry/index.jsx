import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { ListItemIcon } from 'material-ui/List';
import {
  SideNavIcon,
  SideNavText,
  NavItem,
} from './styles';

const SideNavEntry = ({ onClick, link, icon, label, className }) => {
  const item = (
    <NavItem className={className} onClick={onClick}>
      <ListItemIcon><SideNavIcon className={`fa fa-${icon}`} /></ListItemIcon>
      <SideNavText disableTypography inset primary={label} />
    </NavItem>
  );

  if (link) {
    return <Link to={link}>{item}</Link>;
  }

  return item;
};

SideNavEntry.propTypes = {
  onClick: PropTypes.func.isRequired,
  link: PropTypes.string,
  icon: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  className: PropTypes.string,
};

SideNavEntry.defaultProps = {
  className: '',
  link: '',
};

export default SideNavEntry;
