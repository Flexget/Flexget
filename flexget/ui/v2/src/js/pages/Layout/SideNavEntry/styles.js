import styled, { css } from 'react-emotion';
import theme from 'theme';
import Icon from 'material-ui/Icon';
import { ListItem, ListItemText } from 'material-ui/List';

const colorClass = css`color: ${theme.palette.secondary[200]};`;

export const SideNavIcon = styled(Icon)`
  ${colorClass};
`;

export const SideNavText = styled(ListItemText)`
  ${colorClass};
`;

export const NavItem = styled(ListItem)`
  border-left: 3px solid transparent;
  cursor: pointer;

  &:hover {
    border-left: 3px solid ${theme.palette.primary[500]};
  }
`;
