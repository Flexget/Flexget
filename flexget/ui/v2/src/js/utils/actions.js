import { LOADING_STATUS } from 'store/status/actions';

export const action = (type, payload, {
  message,
} = {}) => ({
  type,
  payload,
  error: (payload instanceof Error),
  meta: {
    message,
  },
});

export const request = (type, payload = {}) => ({
  type: LOADING_STATUS,
  payload,
  meta: { type },
});

export const requesting = type => act => (
  act.type === LOADING_STATUS &&
    act.meta.type === type
);
