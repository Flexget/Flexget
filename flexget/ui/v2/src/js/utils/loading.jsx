import Loadable from 'react-loadable';
import Loading from 'common/Loader';

export function createAsyncComponent(loader) {
  return Loadable({
    loading: Loading,
    loader,
  });
}
