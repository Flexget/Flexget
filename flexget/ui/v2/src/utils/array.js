export function groupBy(array, key, transform = f => f) {
  return array.reduce((obj, curr) => {
    const modKey = transform(curr[key]);
    return {
      ...obj,
      [modKey]: obj[modKey] ? [...obj[modKey], curr] : [curr],
    };
  }, {});
}
