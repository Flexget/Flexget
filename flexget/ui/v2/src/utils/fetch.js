function status(response) {
  return response.json().then((data) => {
    if (response.status >= 200 && response.status < 300) {
      return { data, headers: response.headers };
    }
    const err = new Error(data.message);
    err.status = response.status;
    throw err;
  });
}

function request(resource, method, body) {
  const contentType = method === 'get' ? {} : { 'Content-Type': 'application/json' };
  return fetch(`/api${resource}`, {
    method,
    headers: {
      Accept: 'application/json',
      ...contentType,
    },
    credentials: 'same-origin',
    body: JSON.stringify(body),
  })
    .then(status);
}

export function get(resource) {
  return request(resource, 'get');
}

export function post(resource, body) {
  return request(resource, 'post', body);
}

export function put(resource, body) {
  return request(resource, 'put', body);
}

export function del(resource) {
  return request(resource, 'delete');
}
