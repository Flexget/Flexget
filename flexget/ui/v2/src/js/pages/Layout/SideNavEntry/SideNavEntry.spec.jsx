import React from 'react';
import renderer from 'react-test-renderer';
import SideNavEntry from 'pages/Layout/SideNavEntry';
import { themed, router } from 'utils/tests';

describe('pages/Layout/SideNavEntry', () => {
  it('renders correctly with link', () => {
    const tree = renderer.create(
      router(themed(<SideNavEntry icon="icon" label="Test" link="/test" />)),
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly without link', () => {
    const tree = renderer.create(
      themed(<SideNavEntry icon="icon" label="Test" />),
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
