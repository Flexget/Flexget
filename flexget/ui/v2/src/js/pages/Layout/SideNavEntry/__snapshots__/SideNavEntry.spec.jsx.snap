// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Layout/SideNavEntry renders correctly with link 1`] = `
.glamor-2 {
  border-left: 3px solid transparent;
  cursor: pointer;
}

.glamor-2:hover {
  border-left: 3px solid #ff9800;
}

.glamor-0 {
  color: #b0bec5;
}

<a
  href="/test"
  onClick={[Function]}
>
  <li
    className="MuiListItem-root-1 MuiListItem-gutters-8 MuiListItem-default-4 glamor-2"
    disabled={false}
    onClick={undefined}
  >
    <span
      aria-hidden="true"
      className="material-icons MuiIcon-root-12 MuiListItemIcon-root-11 fa fa-icon glamor-0"
    />
    <div
      className="MuiListItemText-root-19 MuiListItemText-inset-20 glamor-0"
    >
      Test
    </div>
  </li>
</a>
`;

exports[`pages/Layout/SideNavEntry renders correctly without link 1`] = `
.glamor-2 {
  border-left: 3px solid transparent;
  cursor: pointer;
}

.glamor-2:hover {
  border-left: 3px solid #ff9800;
}

.glamor-0 {
  color: #b0bec5;
}

<li
  className="MuiListItem-root-1 MuiListItem-gutters-8 MuiListItem-default-4 glamor-2"
  disabled={false}
  onClick={undefined}
>
  <span
    aria-hidden="true"
    className="material-icons MuiIcon-root-12 MuiListItemIcon-root-11 fa fa-icon glamor-0"
  />
  <div
    className="MuiListItemText-root-19 MuiListItemText-inset-20 glamor-0"
  >
    Test
  </div>
</li>
`;
